plugins/fishlim/fish.c

```diff
@@ -60,135 +65,335 @@ static const signed char fish_unbase64[256] = {
+        memcpy(&iv_ciphertext[8], ciphertext, *ciphertext_len);
+        *ciphertext_len += 8;
+
+        g_free(ciphertext);
+        g_free(iv);
+
+        return (char *) iv_ciphertext;
+    } else {
+        return (char *) ciphertext;
+    }
+}
+
+char *fish_encrypt(const char *key, size_t keylen, const char *message, size_t message_len, int mode) {
+    size_t ciphertext_len = 0;
+    char *ciphertext = NULL;
+    char *b64 = NULL;
+
+    if (keylen == 0 || message_len == 0)
+        return NULL;
+
+    ciphertext = fish_cipher(message, message_len, key, keylen, 1, mode, &ciphertext_len);
+
+    if (ciphertext == NULL || ciphertext_len == 0)
+        return NULL;
+
+    switch (mode) {
+        case FISH_CBC_MODE:
+            openssl_base64_encode((const unsigned char *) ciphertext, ciphertext_len, &b64);
+            break;
+
+        case FISH_ECB_MODE:
+            b64 = fish_base64_encode((const char *) ciphertext, ciphertext_len);
+    }
+
+    g_free(ciphertext);
+
+    if (b64 == NULL)
+        return NULL;
+
+    return b64;
+}
+
+char *fish_decrypt(const char *key, size_t keylen, const char *data, int mode) {
+    size_t ciphertext_len = 0;
+    char *ciphertext = NULL;
+    char *plaintext = NULL;
+    char *plaintext_str = NULL;
+
+    if (keylen == 0 || strlen(data) == 0)
+        return NULL;
+
+    switch (mode) {
+        case FISH_CBC_MODE:
+            if (openssl_base64_decode(data, (unsigned char **) &ciphertext, &ciphertext_len) != 0)
+                return NULL;
+            break;
+
+        case FISH_ECB_MODE:
+            ciphertext = fish_base64_decode(data, &ciphertext_len);
+    }
+
+    if (ciphertext == NULL || ciphertext_len == 0)
+        return NULL;
+
+    plaintext = fish_cipher(ciphertext, ciphertext_len, key, keylen, 0, mode, &ciphertext_len);
+    g_free(ciphertext);
+
+    if (ciphertext_len == 0)
+        return NULL;
+
+    plaintext_str = g_malloc0(ciphertext_len + 1);
+    memcpy(plaintext_str, plaintext, ciphertext_len);
+    g_free(plaintext);
+
+    return plaintext_str;
 }
 
+
 /**
  * Encrypts a message (see fish_decrypt). The key is searched for in the
  * key store.
  */
-char *fish_encrypt_for_nick(const char *nick, const char *data) {
+char *fish_encrypt_for_nick(const char *nick, const char *data, int *omode) {
     char *key;
-    char *encrypted;
+    char *encrypted, *encrypted_cbc = NULL;
+    int mode;
+    int encrypted_len = 0;
 
     /* Look for key */
-    key = keystore_get_key(nick);
+    key = keystore_get_key(nick, &mode);
     if (!key) return NULL;
-    
+
+    *omode = mode;
+
     /* Encrypt */
-    encrypted = fish_encrypt(key, strlen(key), data);
-    
+    encrypted = fish_encrypt(key, strlen(key), data, strlen(data), mode);
+
     g_free(key);
-    return encrypted;
+
+    if (encrypted == NULL || mode == FISH_ECB_MODE)
+        return encrypted;
+
+    /* Add '*' for CBC */
+    encrypted_len = strlen(encrypted);
+    encrypted_cbc = g_malloc0(encrypted_len + 2);
```

Comment:
`g_new()`

(I'm not sure the reason you `0`'d it either since its immediately overwritten.)

Reply:
The `0`'d  is for at \0 at end, but anyway i think that is better use the gstring functions

Reply:
If you are filling it up yourself you can `0` it yourself. No point in zeroing the whole thing.

Reply:
Well, I did not think it that way, it has less work yes

Reply:
Oki, better g_strdup_printf here